### PR TITLE
migrate eliza.message system data over to shoebox.keep_event

### PR DESCRIPTION
**Pull request by camhashemi**
@ryanpbrewster @leogrim

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 1]
```diff
@@ -278,6 +278,7 @@ object Eliza extends Service {
 
     def getCrossServiceKeepActivity = ServiceRoute(POST, "/internal/eliza/getCrossServiceKeepActivity")
     def saveKeepEvent = ServiceRoute(POST, "/internal/eliza/saveKeepEvent")
+    def pageSystemMessages(fromId: Id[Message], pageSize: Int) = ServiceRoute(GET, "/internal/eliza/pageSystemMessages", Param("fromId", fromId.id), Param("pageSize", pageSize))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala:281]: `limit` may be more idiomatic

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageRepo.scala [hunk 2]
```diff
@@ -210,6 +213,11 @@ class MessageRepoImpl @Inject() (
     rows.filter(msg => msg.keepId.inSet(keepIds) && msg.seq > seq).sortBy(_.seq).list
   }
 
+  def pageSystemMessagesInclusive(fromId: Id[ElizaMessage], pageSize: Int)(implicit session: RSession): Seq[ElizaMessage] = {
+    val fromTime = activeRows.filter(_.id === fromId).map(_.createdAt).list.head
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageRepo.scala:217]: You may run into issues with this throwing exceptions if you pass in a fake id. I'd either make it `firstOption` here, or have `fromId: Option[Id[ElizaMessage]]`.

File: kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala [hunk 3]
```diff
@@ -375,4 +377,32 @@ class AdminBookmarksController @Inject() (
...
+                keepId = msg.keep,
+                eventData = eventData,
+                eventTime = msg.sentAt,
+                source = KeepEventSourceKind.fromMessageSource(msg.source)
+              )
+              db.readWrite(implicit s => keepEventRepo.save(event))
+            }
+          }
+          startWithMessage = lastMessage.id
+          false
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala]: github isn't showing me the updated version of this code, but we need to handle the case when `msgs.size < pageSize` and we want to persist the entire page (don't pop the last message for pagination). this logic needs to be sound, we don't have a unique index on `keep_event` to intern on, so need to ensure there are no duplicated events before using the table.

I was thinking to add a unique key `message_id` that references `eliza.message.id` to ensure safety.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageRepo.scala [hunk 4]
```diff
@@ -210,6 +213,11 @@ class MessageRepoImpl @Inject() (
...
+    activeRows.filter(msg => msg.auxData.isDefined && msg.createdAt >= fromTime).sortBy(_.id asc).take(pageSize).list
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageRepo.scala:219]: I don't think you want to filter by time here. I think `msg.id >= fromId` or something

File: kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala [hunk 5]
```diff
@@ -375,4 +377,32 @@ class AdminBookmarksController @Inject() (
     }
     Ok.chunked(enum)
   }
+
+  def backfillKeepEventRepo(pageSize: Int, dryRun: Boolean) = AdminUserAction { implicit request =>
+    var startWithMessage = Id[Message](53387) // select min(id) from eliza.message where aux_data is not null;
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala:382]: oh, clever

**Comment by ryanpbrewster** [comment]
Just to reiterate what we said in meatspace:

I think exclusive paging-by-id is simpler and more robust. You can even keep `fromId` mandatory and pass in `Id(0)` to start. The repo query is just going to look like
`activeRows.filter(_.id > fromId).sortBy(_.id asc).take(limit).list`

Although it is technically correct to stop once `retrievedMessages.length < pageSize`, I usually find it to be safer to stop only once `retrievedMessages.isEmpty`. The worst-case is one additional call, and it makes you more robust.

**Comment by ryanpbrewster** [comment]
I'm so pumped for this. `KeepEvent` will be transactional and robust soon, and (although admittedly complex) I think that consistently paging through the distributed messages/events is straightforward. :boom:

File: kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala [hunk 6]
```diff
@@ -375,4 +377,32 @@ class AdminBookmarksController @Inject() (
...
+              db.readWrite(implicit s => keepEventRepo.save(event))
+            }
+          }
+          startWithMessage = lastMessage.id
+          msgs.size < pageSize
+        }
+      }
+    }
+    NoContent
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/admin/AdminBookmarkController.scala:407]: I would urge you to use a chunked response here, but if you're confident then it's kind of superfluous. I just like them :smirk_cat: .

**Comment by camhashemi** [comment]
github isn't updating changes, but thanks @ryanpbrewster

**Pull request merged**